File: apps/web/src/types/crm.ts
export type LeadStatus = "new" | "contacted" | "qualified" | "converted" | "lost";
export type DealStage =
  | "prospecting"
  | "qualification"
  | "proposal"
  | "negotiation"
  | "closed_won"
  | "closed_lost";
export type TaskStatus = "pending" | "in_progress" | "completed" | "cancelled";
export type TaskPriority = "low" | "medium" | "high";

export interface Company {
  id: string;
  user_id: string;
  name: string;
  domain: string | null;
  industry: string | null;
  size: string | null;
  website: string | null;
  phone: string | null;
  address: string | null;
  created_at: string;
  updated_at: string;
}

export interface Contact {
  id: string;
  user_id: string;
  company_id: string | null;
  first_name: string;
  last_name: string;
  email: string | null;
  phone: string | null;
  title: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface Lead {
  id: string;
  user_id: string;
  company_id: string | null;
  contact_id: string | null;
  title: string;
  source: string | null;
  status: LeadStatus;
  score: number | null;
  value: number | null;
  email: string | null;
  phone: string | null;
  notes_summary: string | null;
  created_at: string;
  updated_at: string;
}

export interface Deal {
  id: string;
  user_id: string;
  company_id: string | null;
  contact_id: string | null;
  name: string;
  stage: DealStage;
  amount: number | null;
  currency: string;
  probability: number | null;
  close_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  due_date: string | null;
  related_type: string | null;
  related_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface Note {
  id: string;
  user_id: string;
  body: string;
  related_type: string;
  related_id: string;
  created_at: string;
  updated_at: string;
}

export type CrmEntity = "leads" | "contacts" | "companies" | "deals" | "tasks" | "notes";

export const CRM_ENTITIES: {
  key: CrmEntity;
  label: string;
  singular: string;
  href: string;
}[] = [
  { key: "leads", label: "Leads", singular: "Lead", href: "/dashboard/crm/leads" },
  { key: "contacts", label: "Contacts", singular: "Contact", href: "/dashboard/crm/contacts" },
  { key: "companies", label: "Companies", singular: "Company", href: "/dashboard/crm/companies" },
  { key: "deals", label: "Deals", singular: "Deal", href: "/dashboard/crm/deals" },
  { key: "tasks", label: "Tasks", singular: "Task", href: "/dashboard/crm/tasks" },
  { key: "notes", label: "Notes", singular: "Note", href: "/dashboard/crm/notes" },
];

File: apps/api/app/schemas/crm.py
import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.crm import DealStage, LeadStatus, TaskPriority, TaskStatus


class CompanyBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    domain: str | None = None
    industry: str | None = None
    size: str | None = None
    website: str | None = None
    phone: str | None = None
    address: str | None = None


class CompanyCreate(CompanyBase):
    pass


class CompanyUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    domain: str | None = None
    industry: str | None = None
    size: str | None = None
    website: str | None = None
    phone: str | None = None
    address: str | None = None


class CompanyRead(CompanyBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ContactBase(BaseModel):
    first_name: str = Field(min_length=1, max_length=128)
    last_name: str = Field(min_length=1, max_length=128)
    email: str | None = None
    phone: str | None = None
    title: str | None = None
    status: str = "active"
    company_id: uuid.UUID | None = None


class ContactCreate(ContactBase):
    pass


class ContactUpdate(BaseModel):
    first_name: str | None = Field(default=None, min_length=1, max_length=128)
    last_name: str | None = Field(default=None, min_length=1, max_length=128)
    email: str | None = None
    phone: str | None = None
    title: str | None = None
    status: str | None = None
    company_id: uuid.UUID | None = None


class ContactRead(ContactBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class LeadBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    source: str | None = None
    status: LeadStatus = LeadStatus.NEW
    score: int | None = None
    value: float | None = None
    email: str | None = None
    phone: str | None = None
    notes_summary: str | None = None
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class LeadCreate(LeadBase):
    pass


class LeadUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    source: str | None = None
    status: LeadStatus | None = None
    score: int | None = None
    value: float | None = None
    email: str | None = None
    phone: str | None = None
    notes_summary: str | None = None
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class LeadRead(LeadBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class DealBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    stage: DealStage = DealStage.PROSPECTING
    amount: float | None = None
    currency: str = "USD"
    probability: int | None = None
    close_date: date | None = None
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class DealCreate(DealBase):
    pass


class DealUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    stage: DealStage | None = None
    amount: float | None = None
    currency: str | None = None
    probability: int | None = None
    close_date: date | None = None
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class DealRead(DealBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class TaskBase(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    due_date: date | None = None
    related_type: str | None = None
    related_id: uuid.UUID | None = None


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    status: TaskStatus | None = None
    priority: TaskPriority | None = None
    due_date: date | None = None
    related_type: str | None = None
    related_id: uuid.UUID | None = None


class TaskRead(TaskBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class NoteBase(BaseModel):
    body: str = Field(min_length=1)
    related_type: str = Field(min_length=1, max_length=32)
    related_id: uuid.UUID


class NoteCreate(NoteBase):
    pass


class NoteUpdate(BaseModel):
    body: str | None = Field(default=None, min_length=1)


class NoteRead(NoteBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class CrmActionRequest(BaseModel):
    entity: str
    action: str = "create"
    record_id: uuid.UUID | None = None
    fields: dict[str, Any] = Field(default_factory=dict)
